src/librbd/librbd.cc

```diff
@@ -6140,36 +6140,53 @@ extern "C" int rbd_aio_write2(rbd_image_t image, uint64_t off, size_t len,
   return 0;
 }
 
-extern "C" int rbd_aio_writev(rbd_image_t image, const struct iovec *iov,
-                              int iovcnt, uint64_t off, rbd_completion_t c)
+static int get_iovec_length(const struct iovec *iov, int iovcnt, ssize_t &len)
 {
-  librbd::ImageCtx *ictx = (librbd::ImageCtx *)image;
-  librbd::RBD::AioCompletion *comp = (librbd::RBD::AioCompletion *)c;
-
-  // convert the scatter list into a bufferlist
-  auto aio_completion = get_aio_completion(comp);
-  ssize_t len = 0;
-  bufferlist bl;
+  len = 0;
   for (int i = 0; i < iovcnt; ++i) {
     const struct iovec &io = iov[i];
-    len += io.iov_len;
-    if (len < 0) {
-      break;
+    if (__builtin_add_overflow(len, io.iov_len, &len)) {
```

Comment:
I'm a bit concerned about using compiler builtins for something like this--even though clang appears to implement this;  I would imagine the visual c++ does not, for example

Reply:
It looks like builtins are used all over the code-base. What are the requirements for them to be used? I can check if it compiles with visual c++.
IMO it is cleaner than checking for all the overflow conditions manually. I mean this is why there was a bug in the first place.

Reply:
Alternatively, we could use boost [safe_numerics](https://www.boost.org/doc/libs/1_78_0/libs/safe_numerics/doc/html/tutorial/2.html)?

Reply:
I changed it to now manually check. __builtin_add_overflow does not seem to be supported by visual c++